### ISSUE TITLE
Error in Random forest output values

### ISSUE THREAD

**Issue by aribiere** [davisking/dlib#2095]
## Expected Behavior

I use a `dlib::random_forest_regression_trainer<dlib::dense_feature_extractor>` to create a RF model. It is trained on vectors of 4 variables with nTrees = 20 and Feature subsampling fraction = 1.0. There are more than 100k samples.

Target values are between 0 and 1.2.

I train it with
`auto df = trainer.train(samples, labels);`
I use it with:
`double res = df(samples);`

I think that Random Forest model should predict values inside (or near) the ones used to train the model. But I may misuse it...
## Current Behavior

I've got *VERY* different values when input values differ a litte. Also I get values out of the range of target values...

For example:
```
Input : 0.584699;0.666667;0.127273;0.7859
model output : -3.4313
Input:
0.584699;0.666667;0.127273;0.7858
model output: 0.922213
```
Here the last input's values are very close but the result is *VERY* different.

Here we have negative values but we do not have any negative values in training set...
## Steps to Reproduce

Actually it depends on the model. I can give you a trained model and data to reproduce this if you want.

But it may be a normal behaviour...

Thanks for your help.

* **Version**: 
19.18
* **Where did you get dlib**: 
dlib.net
* **Platform**: 
```bash
$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 9.12 (stretch)
Release:        9.12
Codename:       stretch
```
* **Compiler**: 
```bash
$ g++ --version
g++ (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
```

**Comment by davisking** [comment]
The outputs should be in the range of the training labels.  Double check that you are feeding training the data you think you are.  You probably just have it the wrong data.

**Comment by aribiere** [comment]
Actually I checked input data and I do not have any negative value in it...

**Comment by aribiere** [comment]
Also, I get a Dlib version from Github (7fe355e0ec3da0740c9a2213a2b24a8a01c14fba) and results are a bit different.

I rebuild RF's model but I still have some negative results but less than before. Now for the values given above, I've got:
```
Input : 0.584699;0.666667;0.127273;0.7859
model output : 0.844302
Input:
0.584699;0.666667;0.127273;0.7858
model output: 0.844302
```
which is fine.

But in some cases I've got:
```
Input:
0.584699;0.666667;0.130435;0.792057
Output:
0.841824
Input:
0.584699;0.666667;0.130435;0.794057
Output:
-9.81142
```

**Comment by davisking** [comment]
You need to provide a complete and minimal example that reproduces the problem.

**Comment by aribiere** [comment]
Here is some minimal sample.

A `create_rf_models.cpp` file:
```cpp
#include <fstream>
#include <string>
#include <vector>
#include <cmath>

#include <dlib/random_forest.h>
#include <dlib/svm.h>
#include <dlib/statistics.h>

typedef dlib::matrix<double, 0, 1> sample_type;

int main(int argc, char *argv[]) {

  std::string filePath("data.csv");

  dlib::matrix<double> train_mat;
  std::ifstream ifs(filePath);
  ifs >> train_mat;
  ifs.close();

  dlib::matrix<double> samp_mat, feat;
  samp_mat = dlib::colm(train_mat, dlib::range(0, 3));
  feat = dlib::colm(train_mat, 4);
  std::vector<double> labels(feat.begin(), feat.end());
  std::vector<sample_type> samples;
  for (int i = 0; i < train_mat.nr(); ++i) {
    samples.push_back(dlib::trans(dlib::rowm(samp_mat, i)));
  }

  const unsigned int nTrees = 20;
  const unsigned int mTry = 4;
  std::string file_model("model");

  dlib::random_forest_regression_trainer<dlib::dense_feature_extractor> trainer;
  trainer.set_num_trees(nTrees);
  trainer.set_feature_subsampling_fraction(mTry/static_cast<float>(samples[0].size()));
  trainer.be_verbose();

  auto df = trainer.train(samples, labels);

  // Save model
  dlib::serialize(file_model.c_str()) << df;

  // Read test data
  std::string testPath("test.csv");
  dlib::matrix<double> test_mat;
  std::ifstream test_ifs(testPath);
  test_ifs >> test_mat;
  test_ifs.close();
  for (int i = 0; i < test_mat.nr(); ++i) {
    double res = df(dlib::rowm(test_mat, i));
    std::cout << res << std::endl;
  }
  return 0;
}
```


Once compiled it takes in the same repository `data.csv` and `test.csv`.
[data.csv.gz](https://github.com/davisking/dlib/files/4780926/data.csv.gz)
[test.csv.gz](https://github.com/davisking/dlib/files/4780927/test.csv.gz)

`data.csv` contains the training data. The last column is the one to predict.

When I run it I've got:
```bash
$ bin/create_rf_models
Time remaining: 0 seconds.
0.844302
0.844302
0.811747
0.811747
0.844102
0.844102
0.844302
0.844302
0.783016
0.783016
0.782952
0.782952
0.844302
0.844302
-9.82506
0.783016
-9.82579
0.782283
-9.80779
0.854981
-9.81142
0.841824
```

Negative values shouldn't be there.

**Comment by davisking** [comment]
Thanks for reporting this.  I just took a look and this is a bug in dlib.  I just pushed the fix to github, so it should all work now as expected.

**Comment by aribiere** [comment]
Thanks ! It works fine for us.

Should I close the issue ?

**Comment by davisking** [comment]
Awesome :)
